File: cli/src/derivation.ts
const HARDENED_OFFSET = 0x8000_0000;

export function parseDerivationPath(path: string): number[] {
  if (!path.startsWith("m/")) {
    throw new Error(`Invalid derivation path: ${path}`);
  }

  const parts = path
    .slice(2)
    .split("/")
    .filter((part) => part.length > 0);

  if (parts.length === 0 || parts.length > 5) {
    throw new Error(`Unsupported derivation path length: ${path}`);
  }

  const segments = parts.map((part) => {
    const hardened = part.endsWith("'");
    const valuePart = hardened ? part.slice(0, -1) : part;
    if (!/^\d+$/.test(valuePart)) {
      throw new Error(`Invalid derivation segment: ${part}`);
    }

    const value = Number.parseInt(valuePart, 10);
    if (!Number.isSafeInteger(value) || value < 0 || value >= HARDENED_OFFSET) {
      throw new Error(`Invalid derivation value: ${part}`);
    }

    return hardened ? (value | HARDENED_OFFSET) >>> 0 : value >>> 0;
  });

  if (
    segments[0] !== ((44 | HARDENED_OFFSET) >>> 0) ||
    segments[1] !== ((501 | HARDENED_OFFSET) >>> 0)
  ) {
    throw new Error("Only Solana derivation paths under m/44'/501' are supported");
  }

  return segments;
}

export function serializeDerivationPath(segments: readonly number[]): Buffer {
  if (segments.length === 0 || segments.length > 5) {
    throw new Error("Derivation path must contain 1-5 segments");
  }

  const out = Buffer.alloc(1 + segments.length * 4);
  out[0] = segments.length;
  for (const [index, segment] of segments.entries()) {
    out.writeUInt32BE(segment >>> 0, 1 + index * 4);
  }
  return out;
}

export function formatDerivationPath(segments: readonly number[]): string {
  const formatted = segments.map((segment) => {
    const hardened = (segment & HARDENED_OFFSET) !== 0;
    const value = segment & ~HARDENED_OFFSET;
    return hardened ? `${value}'` : `${value}`;
  });
  return `m/${formatted.join("/")}`;
}

File: cli/src/derivation.rs
use anyhow::{anyhow, bail, Result};

const HARDENED_OFFSET: u32 = 0x8000_0000;

pub fn parse_derivation_path(path: &str) -> Result<Vec<u32>> {
    if !path.starts_with("m/") {
        bail!("invalid derivation path: {path}");
    }

    let parts = path
        .trim_start_matches("m/")
        .split('/')
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>();

    if parts.is_empty() || parts.len() > 5 {
        bail!("unsupported derivation path length: {path}");
    }

    let mut segments = Vec::with_capacity(parts.len());
    for part in parts {
        let hardened = part.ends_with('\'');
        let value_part = if hardened {
            &part[..part.len() - 1]
        } else {
            part
        };

        let value = value_part
            .parse::<u32>()
            .map_err(|_| anyhow!("invalid derivation segment: {part}"))?;
        if value >= HARDENED_OFFSET {
            bail!("invalid derivation value: {part}");
        }

        segments.push(if hardened {
            value | HARDENED_OFFSET
        } else {
            value
        });
    }

    if segments.len() < 2
        || segments[0] != (44 | HARDENED_OFFSET)
        || segments[1] != (501 | HARDENED_OFFSET)
    {
        bail!("only Solana derivation paths under m/44'/501' are supported");
    }

    Ok(segments)
}

pub fn serialize_derivation_path(segments: &[u32]) -> Result<Vec<u8>> {
    if segments.is_empty() || segments.len() > 5 {
        bail!("derivation path must contain 1-5 segments");
    }

    let mut out = Vec::with_capacity(1 + segments.len() * 4);
    out.push(segments.len() as u8);
    for segment in segments {
        out.extend_from_slice(&segment.to_be_bytes());
    }
    Ok(out)
}

pub fn format_derivation_path(segments: &[u32]) -> String {
    let parts = segments
        .iter()
        .map(|segment| {
            let hardened = (segment & HARDENED_OFFSET) != 0;
            let value = segment & !HARDENED_OFFSET;
            if hardened {
                format!("{value}'")
            } else {
                value.to_string()
            }
        })
        .collect::<Vec<_>>();
    format!("m/{}", parts.join("/"))
}

#[cfg(test)]
mod tests {
    use super::{format_derivation_path, parse_derivation_path, serialize_derivation_path};

    #[test]
    fn parses_solana_path() {
        let path = parse_derivation_path("m/44'/501'/0'/0'").unwrap();
        assert_eq!(
            path,
            vec![0x8000_002c, 0x8000_01f5, 0x8000_0000, 0x8000_0000]
        );
        assert_eq!(format_derivation_path(&path), "m/44'/501'/0'/0'");
    }

    #[test]
    fn serializes_path() {
        let encoded = serialize_derivation_path(&[0x8000_002c, 0x8000_01f5]).unwrap();
        assert_eq!(hex::encode(encoded), "028000002c800001f5");
    }
}
